### ISSUE TITLE
read error on connection

### ISSUE THREAD

**Issue by iliaal** [phpredis/phpredis#1543]
Started seeing quite a few of "read error on connection" errors we are using PHP 7.2.16 with php-redis v 4.2.0. The connections are being made to redislabs redis instance. 

Any suggestions on what maybe the issue here?

**Comment by michael-grunder** [comment]
This is usually related to timeouts.  You can try increasing them to see if the issue goes away.

**Comment by iliaal** [comment]
We are currently using a 5s timeout, does that seem like too short of a value?

**Comment by michael-grunder** [comment]
I can't really say whether that's an appropriate timeout since I don't know anything about what you're doing with Redis.

But the `read error on connection` exception is thrown when Redis doesn't reply within the timeout (in your case 5s).  I don't have any experience with using RedisLabs instances but in general, this can happen if you're issuing commands in Redis that take a long time to complete.

Have you tried looking at the `SLOWLOG` output for the instance?

You should be able to get the output directly by issuing a `SLOWLOG GET` command on the node, but it also looks like they offer a [niceer view](https://docs.redislabs.com/latest/rs/administering/logging/redis-slow-log/) as well.

**Comment by yancanchen** [comment]
I also encountered this situation.
1.Code is running in swoole server or fpm. 
2.It's nothing output at the slowlog.

**Comment by TysonAndre** [comment]
The error message could be clearer to avoid confusion? E.g. `(probably timed out)`, `(after %.3f seconds)`, `(no data received)`

Right now, people seeing the error message could think it was an issue with the redis server/proxy, a protocol issue, etc.

```c
         if (redis_sock->port <= 0) {
-            spprintf(&errmsg, 0, "read error on connection to %s", ZSTR_VAL(redis_sock->host));
+            spprintf(&errmsg, 0, "read error on connection to %s (probably timed out)", ZSTR_VAL(redis_sock->host));
        } else {
-            spprintf(&errmsg, 0, "read error on connection to %s:%d", ZSTR_VAL(redis_sock->host), redis_sock->port);
+            spprintf(&errmsg, 0, "read error on connection to %s:%d (probably timed out)", ZSTR_VAL(redis_sock->host), redis_sock->port);
        }
```

**Comment by michael-grunder** [comment]
Unfortunately we can't change it until a major release. I'm not opposed to making it better for 6.0 though.

**Comment by dvdknaap** [comment]
I got the same issue after updating to the latest version.
It's not a connection timeout time on my end because i already changed it in my php ini to unlimited but still get the error.

When i do the redis:flushall on my symfony 4 project i get read error on connection to without any error to debug.

**Comment by michael-grunder** [comment]
Are you sure the timeout is unlimited?  A `read error on connection` error from a `flushall` sounds like timeout (it's possibly a very expensive command).

It's not a phpredis fix but you could use `flushall async` if you're running Redis >= `4.0.0` and see if your problem persists.

**Comment by Venorcis** [comment]
I'm all of a sudden with version 5.2.0 also experiencing a lot of 'read error on connection' as well, mainly on non-localhost connections. We also upgraded to CentOS 8 at the same time so that might be related, but I'm unsure how to proceed with this. Increasing any timeout in PHP does not have any effect (we already had a default socket timeout set for example).

**Comment by michael-grunder** [comment]
Are you connecting with `pconnect`.  I suppose I can see why this would increase with persistent connections because we added a "liveness" check but if you're not using persistent connections I can't think of any reason it would increase off the top of my head.

**Comment by Venorcis** [comment]
> Are you connecting with `pconnect`. I suppose I can see why this would increase with persistent connections because we added a "liveness" check but if you're not using persistent connections I can't think of any reason it would increase off the top of my head.

Thanks for your reply. I'm using regular connect. I tried downgrading to the previous phpredis version butI am still experiencing multiple 'connection resets' per day in some of our worker scripts, which run permanently as a service and thus probably keep the same connection open all the time.

That we started seeing this suddenly is thus probably related to our CentOS 7>8 move instead. It is problematic because the AUTH information is lost after a connection reset, causing the next query after the reset to fail. Would using pconnect actually help in this case? Or do you maybe have any other pointers on why this would happen?

**Comment by Venorcis** [comment]
> I'm all of a sudden with version 5.2.0 also experiencing a lot of 'read error on connection' as well, mainly on non-localhost connections. We also upgraded to CentOS 8 at the same time so that might be related, but I'm unsure how to proceed with this. Increasing any timeout in PHP does not have any effect (we already had a default socket timeout set for example).

This issue seems to have gotten worse with 5.3.0/5.3.1 unfortunately.

Until the upgrade today (we were on 5.2.1 before), it seemed like the 'read error on connection' only happened if a long-running process on a Redis replica was idle for a long time, causing a connection reset when it started up again and then a loss of authentication information (#1757). Now the read error suddenly occurs way more frequently (more towards every 10m instead of every few hours), though still only on the Redis replica server and never in short-lived requests.

**Comment by michael-grunder** [comment]
Read error on connection is only triggered in [redis_sock_gets](https://github.com/phpredis/phpredis/blob/f771ea16b77f39fcca555bec2d952412265197aa/library.c#L3089) and historically is _always_ timeout-related.

If the connection has actually gone bad phpredis tends to fail in [redis_check_eof](https://github.com/phpredis/phpredis/blob/f771ea16b77f39fcca555bec2d952412265197aa/library.c#L302) and you'd see a different error message.

That said we're working around a PHP bug in [redis_check_eof](https://github.com/phpredis/phpredis/blob/f771ea16b77f39fcca555bec2d952412265197aa/library.c#L314) which could have been fixed by now.  Perhaps the workaround is now causing problems.

The hard part is going to be replicating the problem.  In general, what do these long-running scripts do?  I could try putting together a script that runs forever doing generally the same sorts of things.

**Comment by Venorcis** [comment]
> Read error on connection is only triggered in [redis_sock_gets](https://github.com/phpredis/phpredis/blob/f771ea16b77f39fcca555bec2d952412265197aa/library.c#L3089) and historically is _always_ timeout-related.
> 
> If the connection has actually gone bad phpredis tends to fail in [redis_check_eof](https://github.com/phpredis/phpredis/blob/f771ea16b77f39fcca555bec2d952412265197aa/library.c#L302) and you'd see a different error message.
> 
> That said we're working around a PHP bug in [redis_check_eof](https://github.com/phpredis/phpredis/blob/f771ea16b77f39fcca555bec2d952412265197aa/library.c#L314) which could have been fixed by now. Perhaps the workaround is now causing problems.
> 
> The hard part is going to be replicating the problem. In general, what do these long-running scripts do? I could try putting together a script that runs forever doing generally the same sorts of things.

Thank you for the quick response. I understand it'll be hard to replicate this, especially because for us it only seems to happen on replica>master connections (i.e. for writes) and then randomly after long periods of time (15-30 minutes of inactivity probably).

In our case the long-running scripts are job handlers, e.g. for sending push notifications. So these sporadically do a whole bunch of read/writes (mostly hget/hset) and then sleep for a while again. 

Again, the main weird thing is that this used to work perfectly until our CentOS 8 / PhpRedis 5.2 upgrade started causing sporadic connection reset errors in these scripts, which would actually be handled fine if #1757 is fixed, but now the PhpRedis 5.3 causing more frequent read errors (and no other upgrade was performed this time).

**Comment by michael-grunder** [comment]
> Again, the main weird thing is that this used to work perfectly until our CentOS 8 / PhpRedis 5.2 

This makes me think it might actually be that workaround although I won't know until I dig into the issue more closely.

If you don't mind can I get the exact version of CentOS and PHP you're currently running and also the old versions might be useful as I could see if this PHP bug was fixed between the two.  This will also let me run a long-running test script in a container so the environment is as similar as possible.

**Comment by Venorcis** [comment]
> > Again, the main weird thing is that this used to work perfectly until our CentOS 8 / PhpRedis 5.2
> 
> This makes me think it might actually be that workaround although I won't know until I dig into the issue more closely.
> 
> If you don't mind can I get the exact version of CentOS and PHP you're currently running and also the old versions might be useful as I could see if this PHP bug was fixed between the two. This will also let me run a long-running test script in a container so the environment is as similar as possible.

We are currently running the latest CentOS 8.2 (2004) with the latest PHP 7.4.8 (from remi-repo). Before that it was some version of CentOS 7 with some version of PHP 7.3. Perhaps the PHP upgrade is related, but again something also definitely changed between PhpRedis 5.2.1 and 5.3.X.

**Comment by michael-grunder** [comment]
> Perhaps the PHP upgrade is related, but again something also definitely changed between PhpRedis 5.2.1 and 5.3.X.

I believe you that there was a regression between phpredis 5.2.1 and 5.3.x as we reworked how we reauthenticate connections in a way that could certainly cause these symptoms (more calls to `redis_sock_gets` providing more opportunity to trigger `read error on connection`).

It's just possible that the source of the issue is the workaround in question.  I'll certainly let you know if I determine that to be the case :smile:

**Comment by Venorcis** [comment]
~~So one more potentially interesting fact; to work around this issue I decided to close the Redis connection after every job (with `close()`), forcing each new job to use a new PhpRedis instance (that calls `connect(...)` again). However, this strangely does not resolve the issue, still causing read errors on the replicas.~~

EDIT: nevermind, this was caused by another Redis instance that was not closed, sorry.

**Comment by michael-grunder** [comment]
@Venorcis Are your long running jobs subscribed to a pub/sub channel or are they doing something like `BLPOP` in a loop.

Edit:  It would also be nice to have as many `redis.conf` config settings as you could provide. I obviously don't need anything related to security (like auth, etc) but it's more likely I'll be able to cause the problem if I configure the server in a similar way (e.g. with the same timeout, etc)

**Comment by Venorcis** [comment]
> @Venorcis Are your long running jobs subscribed to a pub/sub channel or are they doing something like `BLPOP` in a loop.
> 
> Edit: It would also be nice to have as many `redis.conf` config settings as you could provide. I obviously don't need anything related to security (like auth, etc) but it's more likely I'll be able to cause the problem if I configure the server in a similar way (e.g. with the same timeout, etc)

No, we use neither pub/sub nor blpop. We run Pheanstalk's reserve (see https://github.com/pheanstalk/pheanstalk#consumer--worker) and then on each job that comes in some random read/writes are done (mostly hget and hset like mentioned before). 

I think these are the only non-default settings (on the replica where it occurs):
```
databases 1
replicaof IP PORT
masterauth XYZ
activedefrag yes
```
Do not seem very related/strange right?

As a workaround, when the last job was over 60 seconds ago (just for safety, it can probably be more), we now reconnect to the database (i.e. close the old instance and then create a new one), and since then we have not seen the read error anymore. 

Edit: we call connect like this `$redis->connect($ip, 6379, 1, null, 0, 0, ['auth' => $pass])`

**Comment by michael-grunder** [comment]
Awesome, that's very useful.

I'm glad you were able to work around the issue so it's no longer a pain point but I'm going to try to track it down so the workaround isn't needed.  I'll use that Pheanstalk library as well just so I'm testing the same sort of things.

**Comment by michael-grunder** [comment]
@Venorcis Unfortunately I can't replicate this problem at all.

I went so far as to set up a centos 8.2:2004 docker container running beanstalkd with a producer and consumer PHP script where I can dynamically configure how long we sleep between messages.

Then, I spun up two (primary/replica) Redis instances on an AWS ec2 box and then pointed my consumer at the replica.  When that didn't cause any problems I added another script on the ec2 box that introduced random `DEBUG SLEEP` calls in a loop.

I haven't seen `read error on connection` even once.  Are you certain there is no timeout on your setup?

What does this command give you?
```bash
php -i|grep default_socket_timeout
```

The only way so far that I have been able to trigger a read error on exception is to `DEBUG SLEEP` for more than `60` seconds, which is what `default_socket_timeout` is on my docker container.

It looks like this:

```
Exception: read error on connection to <my-ec2-endpoint>:<my-ec2-port>
PHP Fatal error:  Uncaught Pheanstalk\Exception\ServerException: Job 79 NOT_FOUND: does not exist or is not reserved by client in /root/queue/vendor/pda/pheanstalk/src/Command/ReleaseCommand.php:52
Stack trace:
#0 /root/queue/vendor/pda/pheanstalk/src/Connection.php(114): Pheanstalk\Command\ReleaseCommand->parseResponse()
#1 /root/queue/vendor/pda/pheanstalk/src/Pheanstalk.php(369): Pheanstalk\Connection->dispatchCommand()
#2 /root/queue/vendor/pda/pheanstalk/src/Pheanstalk.php(245): Pheanstalk\Pheanstalk->dispatch()
#3 /root/queue/consumer.php(71): Pheanstalk\Pheanstalk->release()
#4 {main}
```

**Comment by TysonAndre** [comment]
Worth asking: How much data is being used, is SAVE being used (or do regular bgsaves run on replicas?)? During BGSAVE/SAVE, if there's a lot of data, I'd imagine it might hang, but for less than a second in the case of BGSAVE (forking to BGSAVE blocks for longer the more memory redis uses)

From https://redis.io/commands/save

> You almost never want to call SAVE in production environments where it will block all the other clients. Instead usually BGSAVE is used. However in case of issues preventing Redis to create the background saving child (for instance errors in the fork(2) system call), the SAVE command can be a good last resort to perform the dump of the latest dataset.

**Comment by TysonAndre** [comment]
> As a workaround, when the last job was over 60 seconds ago (just for safety, it can probably be more), we now reconnect to the database (i.e. close the old instance and then create a new one), and since then we have not seen the read error anymore.

Immediately after establishing a successful connection, it might be less likely redis is in the middle of saving or completing a long operation (the operation would need to complete before the client could connect), but that's just a guess. Then again, if the 60 seconds is important (e.g. it also has errors with 300 seconds), what I said might not be the case

**Comment by Venorcis** [comment]
Hmm, that's weird, again especially because the behaviour clearly changed for us after updating from 5.2.1 to 5.3.1 without any other changes at that time.

First of all, we do use a socket timeout, but a very high one: `ini_set('default_socket_timeout', 60 * 60 * 24);`
This is to prevent the Pheanstalk connection from dying, and the read error occurred way more often than this.

We do indeed use 'regular bgsave' per the Redis configuration, and our instances are relatively big perhaps.
From redis-cli info: `used_memory_human:3.69G` & `rdb_last_bgsave_time_sec:43`. We never use SAVE.

Edit: the read error occured for us on like 10-20 minute intervals, probably when the worker was sleeping, so testing with the default 60s timeout is probably insufficient to reproduce this.

**Comment by TysonAndre** [comment]
What about OPT_READ_TIMEOUT, or the result of getReadTimeout before you make requests? (e.g. grep for OPT_READ_TIMEOUT in your app/libraries/vendor)

https://github.com/phpredis/phpredis#getreadtimeout

> Mixed Returns the read timeout **(which can be set using setOption and Redis::OPT_READ_TIMEOUT)** or FALSE if we're not connected

**Comment by Venorcis** [comment]
> What about OPT_READ_TIMEOUT, or the result of getReadTimeout before you make requests? (e.g. grep for OPT_READ_TIMEOUT in your app/libraries/vendor)
> 
> https://github.com/phpredis/phpredis#getreadtimeout
> 
> > Mixed Returns the read timeout **(which can be set using setOption and Redis::OPT_READ_TIMEOUT)** or FALSE if we're not connected

Like posted above, we use this constructor: `$redis->connect($ip, 6379, 1, null, 0, 0, ['auth' => $pass])`
That should set the read timeout to 0=infinite right? setOption or OPT_READ_TIMEOUT are never called in our code.

**Comment by yatsukhnenko** [comment]
@Venorcis is it possible to turn off authentication in your environment where you see issue?

**Comment by michael-grunder** [comment]
I don't know if it's related (seems maybe not from scanning through the logic), but it looks like the bug we're working around is now fixed in [all supported versions of PHP](https://git.php.net/?p=php-src.git;a=blobdiff;f=main/streams/xp_socket.c;h=300a58e491d88336980c0cf0e39230148228e0b5;hp=13ed7036826c32dd811a7859ef4198fa6c0c45e8;hb=24f0fe6b1f53e2d610dc27c04b0c6cd9927f4d4a;hpb=b0893344df1946c842d06a8d2c576abb4f7c78e7) so at the very least I'll remove this bit of complexity.

I'll upload this test setup I have to a GitHub repo and maybe @Venorcis can tweak it to be more similar to the production script.  My script doesn't really _do_ anything other than some random HGET/HINCRBY commands

**Comment by Venorcis** [comment]
> @Venorcis is it possible to turn off authentication in your environment where you see issue?

I could try, it's not strictly necessary indeed, but would require quite some work on our end to (temporarily) facilitate. Perhaps an easier approach is to see if I indeed can tweak an example in which the bug is reproducible.

**Comment by vvhungy** [comment]
@Venorcis: what about your redis configs of timeout?

Our redis instances mostly use this:
> timeout 7200
> tcp-keepalive 1200

A short timeout from server may cause client to be read-error and disconnected.

**Comment by Venorcis** [comment]
> @Venorcis: what about your redis configs of timeout?
> 
> Our redis instances mostly use this:
> 
> > timeout 7200
> > tcp-keepalive 1200
> 
> A short timeout from server may cause client to be read-error and disconnected.

Like aforementioned, we only have a few non-default configs. So this should mean that we have `timeout 0` and `tcp-keepalive 300`.

Maybe the tcp-keepalive is related, does the latest phpredis respond to it properly? Crashing every 10m ("to close the connection the double of the time is needed") when there has been no  activity at all (i.e. sleeping php process) sounds like it could potentially be the issue.

**Comment by TysonAndre** [comment]
I'd also have to wonder if it's an issue with a load balancer or firewall.

For example, if a load balancer being connected continues to receives packets, but doesn't send acknowledgement of the packets or reset the connection to indicate that the connection is dead, then the connection may just time out.

Maybe the CentOS 7 server override the default client tcp keepalive setting, and when you rebuilt on CentOS 8, it went back to an extremely high default client keepalive (https://unix.stackexchange.com/questions/519501/default-tcp-keepalive-settings suggests 2 hours or so). (I assume the `tcp-keepalive 1200` applies to what the server is doing to check for dead clients, not the client)

**Comment by Venorcis** [comment]
> I'd also have to wonder if it's an issue with a load balancer or firewall.
> 
> For example, if a load balancer being connected continues to receives packets, but doesn't send acknowledgement of the packets or reset the connection to indicate that the connection is dead, then the connection may just time out.
> 
> Maybe the CentOS 7 server override the default client tcp keepalive setting, and when you rebuilt on CentOS 8, it went back to an extremely high default client keepalive (https://unix.stackexchange.com/questions/519501/default-tcp-keepalive-settings suggests 2 hours or so). (I assume the `tcp-keepalive 1200` applies to what the server is doing to check for dead clients, not the client)

We're directly connecting to the master on LAN, so no load balancing or firewall in between.

And I assume the server keepalive setting is most relevant, i.e. each client has to ACK every keep-alive from the server every 300s by default?

**Comment by michael-grunder** [comment]
@Venorcis I've created a repo to see if we can track this down [the repo is here](https://github.com/michael-grunder/centos-phpredis-beanstalk).  I'm going to invite you as well so you can directly push changes if you find a way to reproduce the issue.

I'll send you host, port, and auth information to my EC2 box out of band (email, or Twitter DM if you prefer) so you can connect to that remote replica if you wish.  Also, if you want to be able to ssh into the box send me your public key in the same message and I'll get it added.

Just please don't start mining bitcoin on the box :joy:

**Comment by pulkitjalan** [comment]
Any updates on this, have you had any luck reproducing this issue?

**Comment by Venorcis** [comment]
> Any updates on this, have you had any luck reproducing this issue?

Sorry, no updates from my side. We now re-connect to Redis if our worker has been idle for longer than a minute, and then only once or twice per month we still get this error.

**Comment by joaomateusferr** [comment]
Hey guys,

I was facing the same "read error on connection" issue when using blPop, in my case, the issue was caused by the connection timeout being less than the blPop timeout, so when the blPop timeout was necessary, the connection with redis was closed due to a timeout before redis was able to answer the blPop result.

**Original Code**

```
<?php

    if(!class_exists('Redis'))
        exit;

    $RedisClient = new Redis();
    $RedisClient->pconnect('127.0.0.1', 6379, 0.5);

    $SampleString = '{"MyData1":"MyData"}';
    $Data = json_decode($SampleString,true);
    $RedisClient->rpush('my-list', json_encode($Data));

    $Result = $RedisClient->blPop('my-list', 1);

    //$Result[0] is the string 'my-list' if the list is not empty

    if(isset($Result[1])){  //check if the list is not empty
        $Data = json_decode($Data,1);
        var_dump($Data);
    }
```

I managed to get around the issue by increasing the connection timeout time to 2 seconds.

**Code Fixed**

```
<?php

    if(!class_exists('Redis'))
        exit;

    $RedisClient = new Redis();
    $RedisClient->pconnect('127.0.0.1', 6379, 2);

    $SampleString = '{"MyData1":"MyData"}';
    $Data = json_decode($SampleString,true);
    $RedisClient->rpush('my-list', json_encode($Data));

    $Result = $RedisClient->blPop('my-list', 1);

    //$Result[0] is the string 'my-list' if the list is not empty

    if(isset($Result[1])){  //check if the list is not empty
        $Data = json_decode($Data,1);
        var_dump($Data);
    }
```
It's worth remembering that is also possible to decrease the blPop timeout but, at least when this text is being written, I couldn't use a float timeout in blPop so to use values smaller than 1 I use the rawCommand method as is pointed by @michael-grunder in the #2157 issue.

**Comment by Samuell1** [comment]
I have same issue, `RedisException: read error on connection to`. Its funny because we run serverless 🚀 it shoudnt return any errors but maybe it scaling and its one second down then it starts erroring?

**Comment by Carlos-CSI** [comment]
I'm also getting the same issue when trying to add a large amount of data to a remote Redis. Everything is set to have 0 timeout but I'm still getting this error.

**Comment by everythings-gonna-be-alright** [comment]
I also get this ```RedisException: read error on connection to```.
Everything works fine if we talk about CLI workers, where we have only one persistent connection.
But FPM, from time to time, got this error. No correlation with higher traffic rate or something else. It just occurs.

I tried different configurations:
On the ElasticCache OSS Redis 7.1 side:
timeout 0 or timeout 300
tcp-keepalive 30 or tcp-keepalive 300

On the FPM side:
pm.max_requests = 30000 or pm.max_requests = 300
default_socket_timeout = 300 or default_socket_timeout = -1

In any combination, this error still occurs.
In my case, errors do not correlate with any specific massive queries. They happen on random queries, even very simple ones.

Environment configuration:
AWS
PHP 8.4 on Alpine Linux 3.21 in Kubernetes with PHPredis 6.2.0
EKS OS Amazon Linux 2023
I use persistent connections and the Redis cluster 7.1 provided by ElasticCache

1-30 errors from millions of queries per day.
This error drives me crazy. Help.

_______
Update:
* I also tried the Debian image instead of Alpine
* Set better sysctl parameters for `tcp_keepalive` on the Kubernetes pod side
* Set pm.max_requests = 0

The first two actions didn't help at all. But I think that setting `pm.max_requests = 0` reduces the number of this type of error. Also, of course, it reduces the number of newly opened connections to Redis, so it correlates with new connections.
But the error didn't disappear, I still have above 100 errors per week.
